docs/models_dev_guide.md

```diff
@@ -0,0 +1,263 @@
+# Models Developer's Guide
+
+This document describes the design of the model management system in ETA and
+best practices for using it.
+
+
+## Motivation
+
+The ETA library relies on a number of large models in order to power its
+analytics. Models are trained on different machines, by different developers,
+and constantly need to be shared. We frequently generate new versions of a
+given model over time as we improve the implementation of the analytic and
+train it on evolving datasets. In addition, models are valuable, so they need
+to be archived and backed up to ensure that they remain secure and readily
+accessible.
+
+The ETA library provides a simple, yet powerful interface that satisifies
+these constraints.
+
+
+## Design
+
+The `eta.core.models` module implements the core interface for managing models
+in ETA. The underlying design principle of the interface is that all models
+are stored in cloud storage and can be easily downloaded to a local ETA
+installation as needed using a simple version-aware naming system.
+
+### Model versions
+
+Each model is identified by a unique string of the form `<name>@<version>`,
+where `<name>` is the canonical name of the model and `<version>` is an
+optional version of the model.
+
+If a model is registered without a version, it can be referenced simply by
+its `<name>`. If a model is registered with a version, it can be referenced
+explicitly using the `<name>@<version>` syntax. The `<name>` syntax can
+also be used with versioned models, in which case the latest version of the
+model (i.e. the largest version number) will be referenced.
+
+#### Models search path
+
+The `models_dirs` field of the ETA config defines the list of directories to
+use when performing all models-related operations. If you add a new models
+directory to your path, you can use the `eta.core.models.init_models_dir`
+function to initialize it for use with the model management interface.
+
+#### Model manifest files
+
+Each directory on the model search path contains a `manifest.json` file that
+describes the models that have been registered in that directory. The manifest
+files are lightweight and are maintained in version control in the repository.
+
+The manifest file contains the names and versions of each model registered in
+its parent directory; it also contain the necessary information to locate
+each model in cloud storage. When a model is downloaded, it will be stored
+in the same directory as its corresponding manifest file.
+
+The following is an example contents of a `manifest.json` file:
+
+```json
+{
+    "models": [
+        {
+            "base_name": "VGG-16",
+            "base_filename": "vgg16_weights.npz",
+            "version": null,
+            "manager": {
```

Comment:
It's fine to stick with this, but I think it is awkward to reference  "manager" field inside of the "models" description.  And, especially including the model file id in the manager field.  Confused why it is organized like this.

Reply:
I suppose you answer this below.  Still noting a bit of confusion, but I get it.

Reply:
`manager` is written like this because it is a `Configurable` class like `FeaturizerConfig` that can construct subclasses of itself on-demand. This allows us to support remote remote storage solutions.

ETA models need to be publicly accessible, but THETA models need to be private, so the information in `manager` depends on the cloud storage strategy. Everything else is local information needed to work with the models on-premises.